apps/heft/src/cli/actions/StartAction.ts

```diff
@@ -1,19 +1,24 @@
 // Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
 // See LICENSE in the project root for license information.
 
-import { HeftActionBase, IHeftActionBaseOptions, ActionHooksBase, IActionDataBase } from './HeftActionBase';
+import { HeftActionBase, IHeftActionBaseOptions, ActionHooksBase, IActionContext } from './HeftActionBase';
 
 /**
  * @public
  */
-export class StartHooks extends ActionHooksBase {}
+export class StartHooks extends ActionHooksBase<IStartActionProperties> {}
 
 /**
  * @public
  */
-export interface IStartActionData extends IActionDataBase<StartHooks> {}
+export interface IStartActionProperties {}
 
-export class StartAction extends HeftActionBase<IStartActionData, StartHooks> {
+/**
+ * @public
+ */
+export interface IStartActionContext extends IActionContext<StartHooks, IStartActionProperties> {}
```

Comment:
The generic `<StartHooks, IStartActionProperties>` has a couple disadvantages:

- This pattern cannot be extended further into a child interface that extends `IStartActionContext`
- The API documentation is abstracted.
- The generic parameters must be specified even when there are no changes to the hooks/properties
- If more inheritable items are added to the context, they will accumulate as generic parameters.

Compare with this:

```ts
export interface IStartActionContext extends IActionContext {
  hooks: StartHooks;  // (redefines base member)
  properties: IStartActionProperties;  // (redefines base member)
}
```

Might be more verbose to write, but for a person reading the API docs, the type of `hooks` and `properties` is much easier to understand.

Reply:
>This pattern cannot be extended further into a child interface that extends IStartActionContext

Anything that extends `IStartActionContext` can add properties

>The API documentation is abstracted.

I view this as a positive - the API documentation exists in one place.

>The generic parameters must be specified even when there are no changes to the hooks/properties

There's no default implementation for either, so I also view this as a positive.

>If more inheritable items are added to the context, they will accumulate as generic parameters.

That's a legitimate concern, so I'd say we should consider this change if/when we get to that point.

That alternate implementation is also far more likely to cause someone to make a mistake. Additionally, having the generics on the base types (and no generics on the child types) helps indicate that the base types used on the base interface properties aren't the ones any consumer of the API should be using.